File: exercises/exercise-1-testing-java/src/main/java/de/emaarco/example/domain/bike/BikeSubscription.java
package de.emaarco.example.domain.bike;

import de.emaarco.example.domain.Email;
import de.emaarco.example.domain.Name;

import java.time.LocalDateTime;
import java.util.UUID;

public record BikeSubscription(
        BikeSubscriptionId id,
        BikeId bikeId,
        Email customerEmail,
        Name customerName,
        LocalDateTime requestDate,
        BikeSubscriptionStatus status,
        Boolean bikeAvailable
) {
    public BikeSubscription(BikeId bikeId, Email customerEmail, Name customerName) {
        this(
                new BikeSubscriptionId(UUID.randomUUID()),
                bikeId,
                customerEmail,
                customerName,
                LocalDateTime.now(),
                BikeSubscriptionStatus.PENDING,
                null
        );
    }

    public BikeSubscription markAvailability(boolean available) {
        return new BikeSubscription(
                id,
                bikeId,
                customerEmail,
                customerName,
                requestDate,
                available ? BikeSubscriptionStatus.AVAILABLE : BikeSubscriptionStatus.NOT_AVAILABLE,
                available
        );
    }

    public BikeSubscription markPaymentReceived() {
        return new BikeSubscription(
                id,
                bikeId,
                customerEmail,
                customerName,
                requestDate,
                BikeSubscriptionStatus.PAID,
                bikeAvailable
        );
    }

    public BikeSubscription markShipped() {
        return new BikeSubscription(
                id,
                bikeId,
                customerEmail,
                customerName,
                requestDate,
                BikeSubscriptionStatus.SHIPPED,
                bikeAvailable
        );
    }

    public BikeSubscription markActive() {
        return new BikeSubscription(
                id,
                bikeId,
                customerEmail,
                customerName,
                requestDate,
                BikeSubscriptionStatus.ACTIVE,
                bikeAvailable
        );
    }

    public BikeSubscription cancel() {
        return new BikeSubscription(
                id,
                bikeId,
                customerEmail,
                customerName,
                requestDate,
                BikeSubscriptionStatus.CANCELED,
                bikeAvailable
        );
    }
}

File: exercises/exercise-1-testing-kotlin/src/main/kotlin/de/emaarco/example/domain/bike/BikeSubscription.kt
package de.emaarco.example.domain.bike

import de.emaarco.example.domain.Email
import de.emaarco.example.domain.Name
import java.time.LocalDateTime

data class BikeSubscription(
    val id: BikeSubscriptionId = BikeSubscriptionId(java.util.UUID.randomUUID()),
    val bikeId: BikeId,
    val customerEmail: Email,
    val customerName: Name,
    val requestDate: LocalDateTime = LocalDateTime.now(),
    val status: BikeSubscriptionStatus = BikeSubscriptionStatus.PENDING,
    val bikeAvailable: Boolean? = null
) {
    fun markAvailability(available: Boolean) = copy(
        bikeAvailable = available,
        status = if (available) BikeSubscriptionStatus.AVAILABLE else BikeSubscriptionStatus.NOT_AVAILABLE
    )

    fun markPaymentReceived() = copy(status = BikeSubscriptionStatus.PAID)

    fun markShipped() = copy(status = BikeSubscriptionStatus.SHIPPED)

    fun markActive() = copy(status = BikeSubscriptionStatus.ACTIVE)

    fun cancel() = copy(status = BikeSubscriptionStatus.CANCELED)
}
